sdk/storage/azdatalake/sas/service.go

```diff
@@ -0,0 +1,407 @@
+		v.ContentLanguage,    // rscl
+		v.ContentType},       // rsct
+		"\n")
+
+	signature, err := exported.ComputeHMACSHA256(sharedKeyCredential, stringToSign)
+	if err != nil {
+		return QueryParameters{}, err
+	}
+
+	p := QueryParameters{
+		// Common SAS parameters
+		version:     v.Version,
+		protocol:    v.Protocol,
+		startTime:   v.StartTime,
+		expiryTime:  v.ExpiryTime,
+		permissions: v.Permissions,
+		ipRange:     v.IPRange,
+
+		// Container/Blob-specific SAS parameters
+		resource:             resource,
+		identifier:           v.Identifier,
+		cacheControl:         v.CacheControl,
+		contentDisposition:   v.ContentDisposition,
+		contentEncoding:      v.ContentEncoding,
+		contentLanguage:      v.ContentLanguage,
+		contentType:          v.ContentType,
+		signedDirectoryDepth: getDirectoryDepth(v.DirectoryPath),
+		authorizedObjectID:   v.AuthorizedObjectID,
+		unauthorizedObjectID: v.UnauthorizedObjectID,
+		correlationID:        v.CorrelationID,
+		// Calculated SAS signature
+		signature: signature,
+	}
+
+	return p, nil
+}
+
+// SignWithUserDelegation uses an account's UserDelegationCredential to sign this signature values to produce the proper SAS query parameters.
+func (v DatalakeSignatureValues) SignWithUserDelegation(userDelegationCredential *UserDelegationCredential) (QueryParameters, error) {
+	if userDelegationCredential == nil {
+		return QueryParameters{}, fmt.Errorf("cannot sign SAS query without User Delegation Key")
+	}
+
+	if v.ExpiryTime.IsZero() || v.Permissions == "" {
+		return QueryParameters{}, errors.New("user delegation SAS is missing at least one of these: ExpiryTime or Permissions")
+	}
+
+	// Parse the resource
+	resource := "c"
+	if v.DirectoryPath != "" {
+		resource = "d"
+		v.FilePath = ""
+	} else if v.FilePath == "" {
+		// do nothing
+	} else {
+		resource = "b"
+	}
+	// make sure the permission characters are in the correct order
+	if resource == "c" {
+		perms, err := parseFilesystemPermissions(v.Permissions)
+		if err != nil {
+			return QueryParameters{}, err
+		}
+		v.Permissions = perms.String()
+	} else {
+		perms, err := parsePathPermissions(v.Permissions)
+		if err != nil {
+			return QueryParameters{}, err
+		}
+		v.Permissions = perms.String()
+	}
+
+	if v.Version == "" {
+		v.Version = Version
+	}
+	startTime, expiryTime := formatTimesForSigning(v.StartTime, v.ExpiryTime)
+
+	udk := exported.GetUDKParams(userDelegationCredential)
+
+	udkStart, udkExpiry := formatTimesForSigning(*udk.SignedStart, *udk.SignedExpiry)
+
+	stringToSign := strings.Join([]string{
+		v.Permissions,
+		startTime,
+		expiryTime,
+		getCanonicalName(exported.GetAccountName(userDelegationCredential), v.FilesystemName, v.FilePath, v.DirectoryPath),
+		*udk.SignedOID,
+		*udk.SignedTID,
+		udkStart,
+		udkExpiry,
+		*udk.SignedService,
+		*udk.SignedVersion,
+		v.AuthorizedObjectID,
+		v.UnauthorizedObjectID,
+		v.CorrelationID,
+		v.IPRange.String(),
+		string(v.Protocol),
+		v.Version,
+		resource,
```

Comment:
same here does encryption scope and snapshot need empty spaces here?

Reply:
same as above